lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -612,98 +794,90 @@ finalizeModel(GeneratedCodeBasicInfo &GCBI,
             // It's a dynamic function call
-            revng_assert(CE->Type == model::FunctionEdgeType::FunctionCall);
+            revng_assert(CE->Type == efa::FunctionEdgeType::FunctionCall);
             CE->Destination = MetaAddress::invalid();
             CE->DynamicFunction = SymbolName.str();
 
-            // The prototype is implicitly the one of the callee
-            revng_assert(not CE->Prototype.isValid());
+            // The prototype must not exist among the ones of the call sites of
+            // the function, it is implicitly the one of the callee.
+            revng_assert(not Function->CallSitePrototypes.count(Block.Start));
           } else if (CE->Destination.isValid()) {
             // It's a simple direct function call
-            revng_assert(CE->Type == model::FunctionEdgeType::FunctionCall);
+            revng_assert(CE->Type == efa::FunctionEdgeType::FunctionCall);
 
-            // The prototype is implicitly the one of the callee
-            revng_assert(not CE->Prototype.isValid());
+            // The prototype must not exist among the ones of the call sites of
+            // the function, it is implicitly the one of the callee.
+            revng_assert(not Function->CallSitePrototypes.count(Block.Start));
           } else {
             // It's an indirect call: forge a new prototype
-            auto Prototype = buildPrototype(GCBI, Binary, Summary, Block);
-            CE->Prototype = Binary.recordNewType(std::move(Prototype));
+            auto Prototype = buildPrototype(Summary, Block);
+            auto TypedPrototype = Binary.recordNewType(std::move(Prototype));
+            auto PrototypeInserter = Function->CallSitePrototypes
+                                       .batch_insert();
+            PrototypeInserter.insert({ Block.Start, TypedPrototype });
           }
         }
       }
     }
 
-    Function.CFG = Summary.CFG;
-    revng_check(Function.verify(true));
+    efa::FunctionMetadata FM(Function->Entry, Summary.CFG);
+    FM.verify(Binary, true);
   }
+
+  revng_check(Binary.verify(true));
 }
 
-static void
-combineCrossCallSites(MetaAddress EntryPC, FunctionSummary &Summary) {
+static void combineCrossCallSites(auto &CallSite, auto &Callee) {
   using namespace ABIAnalyses;
-  using RegState = abi::RegisterState::Values;
+  using RegisterState = abi::RegisterState::Values;
 
-  for (auto &[PC, CallSites] : Summary.ABIResults.CallSites) {
-    if (PC == EntryPC) {
-      for (auto &[FuncArg, CSArg] :
-           zipmap_range(Summary.ABIResults.ArgumentsRegisters,
-                        CallSites.ArgumentsRegisters)) {
-        auto *CSV = FuncArg == nullptr ? CSArg->first : FuncArg->first;
-        auto RSFArg = FuncArg == nullptr ? RegState::Maybe : FuncArg->second;
-        auto RSCSArg = CSArg == nullptr ? RegState::Maybe : CSArg->second;
+  for (auto &[FuncArg, CSArg] :
+       zipmap_range(Callee.ArgumentsRegisters, CallSite.ArgumentsRegisters)) {
```

Comment:
Maybe it's not strictly related to this PR (or more likely I'm just being way too paranoid), but the trauma of having unexpected problems with `zip`-ers is still way too fresh (there's a discussion with @drblallo somewhere in `~libsdg`). What I'm leading towards, maybe use cases like this should actually assert that both `lhs` and `rhs` containers have the same size - just to be double-safe.

Reply:
I'm all for extra assertions! :)